cli/internal/cache/cache_http.go

```diff
@@ -206,56 +207,73 @@ func (cache *httpCache) retrieve(hash string) (bool, []string, int, error) {
 			err = fmt.Errorf("artifact verification failed: artifact tag does not match expected tag %s", expectedTag)
 			return false, nil, 0, err
 		}
+		// We read the old artifactReader, we can close it.
+		// The defered call to artifactReader.Close will pick up the new value
+		_ = artifactReader.Close()
 		// The artifact has been verified and the body can be read and untarred
 		artifactReader = ioutil.NopCloser(bytes.NewReader(b))
 	}
+	files, err := restoreTar(cache.repoRoot, artifactReader)
+	if err != nil {
+		return false, nil, 0, err
+	}
+	return true, files, duration, nil
+}
+
+func restoreTar(root fs.AbsolutePath, reader io.Reader) ([]string, error) {
 	files := []string{}
 	missingLinks := []*tar.Header{}
-	gzr, err := gzip.NewReader(artifactReader)
+	gzr, err := gzip.NewReader(reader)
 	if err != nil {
-		return false, nil, 0, err
+		return nil, err
 	}
-	defer gzr.Close()
+	defer func() { _ = gzr.Close() }()
 	tr := tar.NewReader(gzr)
 	for {
 		hdr, err := tr.Next()
 		if err != nil {
 			if err == io.EOF {
 				for _, link := range missingLinks {
-					err := restoreSymlink(link, true)
+					err := restoreSymlink(root, link, true)
 					if err != nil {
-						return false, nil, 0, err
+						return nil, err
 					}
 				}
 
-				return true, files, duration, nil
+				return files, nil
 			}
-			return false, nil, 0, err
+			return nil, err
 		}
 		files = append(files, hdr.Name)
+		filename := root.Join(hdr.Name)
```

Comment:
is `hdr.Name` always a Unix path? The test has `expectedFiles` as `/` separated, if so, this line implicitly `Clean`s them to system separators.

Reply:
Yes, we convert all paths to posix-style on write into the tar file.

Reply:
Cool, can you add a comment here so that I remember to do it when I make it do the right thing inside of `turbopath`?

Reply:
Traced through this a little more, and I think it gets to our not-great cache abstraction. I don't think I've changed the behavior here, but I also think the return value, which will be posix-style repo-relative paths, is both wrong but also works due to Windows accepting `/`-delimited paths. This is used when a cache hit from remote cache is used to populate the filesystem cache. The return type here *should* match the argument type for files to `cache.Put`. That should be `RepoRelativeSystemPath`.

I'll add a comment, but I think this is best cleaned up when we have the types to clarify what we actually want.